src/game/middlewares/puzzle-solution-middleware.js

```diff
@@ -0,0 +1,41 @@
+import { RECEIVE_PUZZLE } from '../../shared/actions/puzzle-actions';
+import { SET_SELECTOR, setSelection } from '../actions/solution-actions';
+
+const currentPuzzle = {
+    nodes: null,
+};
+
+const puzzleSolutionMiddleware = store => next => action => {
+    if (action.type === RECEIVE_PUZZLE) {
+        currentPuzzle.nodes = createPuzzleNodesTree(action.payload.markup);
+    }
+
+    if (action.type === SET_SELECTOR) {
+        try {
+            const selection = applySelector(action.payload);
+            store.dispatch(setSelection(selection));
+        } catch (err) {
+            store.dispatch(setSelection([]));
+        }
+    }
+
+    let result = next(action);
+    return result;
+};
+
+function createPuzzleNodesTree(markup) {
+    const container = document.createElement('div');
+    container.innerHTML = markup;
+    return container;
+}
+
+function applySelector(selector) {
+    const selectedNodes = currentPuzzle.nodes.querySelectorAll(selector);
```

Comment:
I feel a bit uneasy about this being browser-dependent
There's fairly recent stuff like [this](https://stackoverflow.com/questions/35787373/safari-9-fails-queryselectorall-with-invalid-selector), for example.

Reply:
that's a temporary middleware just to verify everything works as expected. Later this will be replaced with server messages handlers

Reply:
Ah, of course, got it